packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -641,21 +642,22 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
     /**
      * Load the stores from a snapshot and returns the runtime.
      * @param context - Context of the container.
-     * @param registryEntries - Mapping to the stores.
-     * @param requestHandler - Request handlers for the container runtime
-     * @param runtimeOptions - Additional options to be passed to the runtime
-     * @param existing - (optional) When loading from an existing snapshot. Precedes context.existing if provided
-     * @param containerRuntimeCtor - (optional) Constructor to use to create the ContainerRuntime instance. This
-     * allows mixin classes to leverage this method to define their own async initializer.
-     */
-    public static async load(
+     * @param containerRuntimeCtor - (optional) Constructor to use to create the ContainerRuntime instance.
+     * This allows mixin classes to leverage this method to define their own async initializer.
+     * @param existing - Pass 'true' if loading from an existing snapshot.
+     * @param registryEntries - Mapping from data store types to their corresponding factories,
+     * @param containerScop - Scope
+     * @param initializeEntryPoint - Promise that resolves to an object which will act as entryPoint for the Container.
+     * This object should provide all the functionality that the Container is expected to provide to the loader layer.
+     */
+    public static async newLoad(
         context: IContainerContext,
-        registryEntries: NamedFluidDataStoreRegistryEntries,
-        requestHandler?: (request: IRequest, runtime: IContainerRuntime) => Promise<IResponse>,
+        containerRuntimeCtor: typeof ContainerRuntime = ContainerRuntime,
         runtimeOptions: IContainerRuntimeOptions = {},
+        existing: boolean,
+        registryEntries: NamedFluidDataStoreRegistryEntries,
```

Comment:
i would really like if we could get rid of registryEntries as well, and just have it be something the entrypoint supports, specifically IProvideFluidDataStoreRegistry

Reply:
We might be able to do some more progress towards that if I initialize the entryPoint synchronously like @ChumpChief suggested [here](https://github.com/microsoft/FluidFramework/pull/12487#discussion_r1080676488). Because right now the container runtime takes the registry as a field in its constructor, so it's available immediately when you get a runtime instance. If we synchronously initialize the entryPoint in the runtime's constructor, we can get the registry entries out of it and create a new registry inside the constructor (and overwrite the one from the constructor parameter, because it isn't optional, so we'd still need to provide an empty registry at least). And in a breaking change remove the `private readonly registry: IFluidDataStoreRegistry` parameter.

My concern in that scenario is that we're locking when the entryPoint initialization function is called and saying that it only has access to a runtime instance, but that instance has not gone through any *other* (e.g. async) initialization steps. Does this seem like an ok trade-off? The registry feels like something that any further initialization might need, e.g. to instantiate data stores, so I don't think we can delay its availability. Do we let that dictate the lifecycle for the entryPoint? I think I lack the experience to give a solid answer to that.

Reply:
> i would really like if we could get rid of registryEntries as well, and just have it be something the entrypoint supports, specifically IProvideFluidDataStoreRegistry

A different opinion - I believe that currently `entryPoint` is a host-facing API, i.e. everything on it is something the container author intends to provide as API surface to the host.  The registry would not fit this pattern, as it should be private to the container contents.  I would suggest keeping the `entryPoint` dedicated to the single responsibility of being a host-facing API entry point and find some other way to hide the registry if we think that's desirable.

Reply:
I don't disagree with the above, but the other side is that it forces a specific implementation pattern where we force the consumer to implement multiple specialized objects. This keeps the runtime in the middle of everything rather than delegating control to the consumer.

I'm actually curious what @leeviana thinks here, as she has a ton of experience building registries.

Reply:
"Consumer" here meaning the container author, correct? I don't see a problem with asking them for a single composable object per se, but then I think that object would need to be something like

```typescript
interface SingleObjectThatContainerAuthorNeedsToProvide {
  entryPoint: FluidObject,

  // This one could be split in several properties like registries, scope, etc
  configurationForTheRuntime: SomeNewInterfaceThatEncompassesAllCurrentPieces,
}
```

But have something that clearly lets us separate the API-for-the-host piece from the things-FF-needs-to-run-me piece, so we can expose the first one as necessary but not the second one. Would that be in line with the approach you have in mind, @anthony-murphy ?

Reply:
> I don't disagree with the above, but the other side is that it forces a specific implementation pattern where we force the consumer to implement multiple specialized objects.

I think this requirement is a positive.  The container author has two distinct responsibilities they need to think about:
1. Build a great API surface for their host app to consume (write a really good entryPoint)
2. Set up the container internals to provide the necessary internal machinery to make that entryPoint work as intended (e.g. registry).

Regardless of whether these are passed as a single object or two separate objects, the total burden of the container author is the same.  By encouraging separation of the object though, we can encourage container authors to avoid blending the two and spaghetti-ing between host and internal (e.g. If a host can access the registry, they might think it's appropriate to modify, query, etc. that registry unbeknownst to the container.  Which would be terrible).

If/when we get the runtime out of the middle, these will be separate anyway so this aligns with that goal.  The model loading stuff is kind of doing this already by composing the container runtime into the model (thereby hiding the registry and such while exposing the host-specific interface).

Reply:
I don't have the full context here, but registry extensibility has definitely been a part of the FF API that we've had to work around and it would be great to be able to register a provider (fluidobject) pattern that composes registration requirements. Actually, that is exactly the ask here: https://github.com/microsoft/FluidFramework/issues/4193

In essence, we have quite a few pieces of information that dataobjects can expose as part of their entrypoint information (preload/properties of the experience itself) in addition to the factory that FF requires. Today we have to essentially maintain two parallel registry flows from both dataobject developers (have to expose both) and dataobject consumer flows (have to use/compose both). This adds in registry complexity in places where we wouldn't want them to know about registries. I haven't tried to play around with this area since the move to providers to see if there's already something implicit we can cast the internals of NamedFluidDataStoreRegistryEntries to.

Maybe not as directly relevant to the question but two more related thoughts to consider:
* The Aqueduct "subcomponent" registration pattern also has this problem, we've worked around it today by wrapping Aqueduct's factory, but we'd prefer the wrapper to not be needed.
* The registration/factory interface is the single point of breaking API changes between dataobjects and container registrations patterns on differing versions of packages. Moving these to a provider pattern instead of a single strict interface pattern would alleviate this problem (we'd need to be mindful of changing the interfaces still but unrelated changes to different flows won't affect people who don't use it)

Reply:
With regards to host vs component consumption of container logic, from the container perspective we do split those and they do not interact. I would keep in mind the subcomponent registration bullet point though - if you are adding something to the registration pattern, **component developers will consume these objects as well via their registration pattern**s. These we do not want to have host logic be a part of.